File: algorithms/easy/calculate_digit_sum_of_a_string.py
class Solution:
    def digitSum(self, s: str, k: int) -> str:
        def strSum(s):
            return str(sum([int(ch) for ch in s]))

        while len(s) > k:
            t = ''
            i = 0
            while i < len(s):
                t += strSum(s[i:i+k])
                i += k
            s = t
        return s

# Main section
for s, k in [
               ('11111222223', 3),
               ('00000000', 3),
               ('99999999999999999999999999999999999999999999999999999999999999999', 4),
            ]: 
    print(f's, k = {s}, {k}')
    sol = Solution()
    r = sol.digitSum(s, k)
    print(f'r = {r}')
    print('=========================')

File: algorithms/easy/calculate_digit_sum_of_a_string.go
package main

import (
    "fmt"
    "strconv"
)

func min(a, b int) int {
    if a <= b {
        return a
    }
    return b
}

func digitSum(s string, k int) string {
    for len(s) > k {
        tmp := ""
        for i := 0; i < len(s); i = i + k {
            chunk := s[i:min(i+k, len(s))]
            sum := 0
            for _, v := range chunk {
                sum += int(v-'0')
            }
            tmp += strconv.Itoa(sum)
        }
        s = tmp
    }
    return s
}

func main() {
    type Pair struct {
        s  string
        k  int
    }
    for _, v := range []Pair {
                               Pair{"11111222223", 3},
                               Pair{"00000000", 3},
                             }{
        fmt.Printf("s = %s ; k = %d\n", v.s, v.k)
        r := digitSum(v.s, v.k)
        fmt.Printf("r = %s\n", r)
        fmt.Println("===================")
    }
}

